editoast/src/views/v2/path/pathfinding.rs

```diff
@@ -95,117 +94,201 @@ pub async fn post(
         infra_id: *infra_id,
     })
     .await?;
+
     Ok(Json(
-        pathfinding_blocks(conn, &mut redis_conn, core, &infra, &path_input).await?,
+        pathfinding_blocks(conn, &mut redis_conn, core, &infra, path_input).await?,
     ))
 }
 
+/// Pathfinding computation given a path input
 async fn pathfinding_blocks(
     conn: &mut DbConnection,
     redis_conn: &mut RedisConnection,
     core: Arc<CoreClient>,
     infra: &Infra,
-    path_input: &PathfindingInput,
+    path_input: PathfindingInput,
 ) -> Result<PathfindingResult> {
-    // Compute unique hash of PathInput
-    let hash = path_input_hash(infra.id, &infra.version, path_input);
+    let mut path = pathfinding_blocks_batch(conn, redis_conn, core, infra, &[path_input]).await?;
+    Ok(path.pop().unwrap())
+}
+
+/// Pathfinding batch computation given a list of path inputs
+async fn pathfinding_blocks_batch(
+    conn: &mut DbConnection,
+    redis_conn: &mut RedisConnection,
+    core: Arc<CoreClient>,
+    infra: &Infra,
+    pathfinding_inputs: &[PathfindingInput],
+) -> Result<Vec<PathfindingResult>> {
+    // Compute hashes of all path_inputs
+    let hashes: Vec<_> = pathfinding_inputs
+        .iter()
+        .map(|input| path_input_hash(infra.id, &infra.version, input))
+        .collect();
+
     // Try to retrieve the result from Redis
-    let result: Option<PathfindingResult> =
-        redis_conn.json_get_ex(&hash, CACHE_PATH_EXPIRATION).await?;
-    if let Some(pathfinding) = result {
-        info!("Hit cache");
-        return Ok(pathfinding);
-    }
-    // If miss cache:
+    let mut pathfinding_results: Vec<Option<PathfindingResult>> =
+        redis_conn.json_get_bulk(&hashes).await?;
+
+    // Report number of hit cache
+    let nb_hit = pathfinding_results
+        .iter()
+        .filter(|pathfinding_result| pathfinding_result.is_some())
+        .count();
+    info!(
+        nb_hit,
+        nb_miss = pathfinding_inputs.len() - nb_hit,
+        "Hit cache"
+    );
+
+    // Handle miss cache:
     // 1) extract locations from path items
-    let path_items = path_input.clone().path_items;
-    if path_items.len() <= 1 {
-        return Ok(PathfindingResult::NotEnoughPathItems);
-    }
-    let result = extract_location_from_path_items(conn, infra.id, &path_items).await?;
-    let track_offsets = match result {
-        Ok(track_offsets) => track_offsets,
-        Err(e) => return Ok(e.into()),
-    };
-
-    // Check if tracks exist
-    if let Err(pathfinding_result) =
-        check_tracks_from_path_items(conn, infra.id, &track_offsets).await?
+    let mut to_cache = vec![];
+    let mut pathfinding_requests = vec![];
+    let mut pathfinding_requests_index = vec![];
+    for (index, (pathfinding_result, pathfinding_input)) in pathfinding_results
+        .iter_mut()
+        .zip(pathfinding_inputs)
+        .enumerate()
     {
-        return Ok(pathfinding_result);
+        if pathfinding_result.is_some() {
+            continue;
+        }
```

Comment:
This could be moved in a build_pathfinding_request function, return a Result<PathfindingRequest, PathfindingResult>, and the boiler plate of adding some stuff to cache and updating patfinding result would be factorized

Reply:
I messed up my diff selection for this comment, it relates to the body of the main for loop of `pathfinding_blocks_batch`

Reply:
Good catch.